tests/test_symbols.py

```diff
@@ -25,225 +25,224 @@
+
+                # Get the symbol back
+                y, = itertools.islice(new.iter_symbols(), index, index+1)
+
+                self.assertFalse(x.shares_memory(y))
+                self.assertIs(type(x), type(y))
+                self.assertTrue(x.equals(y))
+
+        def test_state_serialization(self):
+            for x in self.generate_symbols():
+                model = x.model
+
+                # without some way to randomly initialize states this is really just
+                # smoke test
+                model.states.resize(1)
+
+                # get the states before serialization
+                states = []
+                for sym in model.iter_symbols():
+                    if hasattr(sym, "state"):
+                        states.append(sym.state())
+                    else:
+                        states.append(None)
+
+                with model.states.to_file() as f:
+                    new = Model.from_file(f)
+
+                for i, sym in enumerate(model.iter_symbols()):
+                    if hasattr(sym, "state"):
+                        np.testing.assert_equal(sym.state(), states[i])
+
+        def test_state_size_smoke(self):
+            for x in self.generate_symbols():
+                self.assertGreaterEqual(x.state_size(), 0)
+
+    class BinaryOpTests(SymbolTests):
+        @abc.abstractmethod
+        def op(self, lhs, rhs):
+            pass
+
+        def symbol_op(self, lhs, rhs):
+            # if the op is different for symbols, allow the override
+            return self.op(lhs, rhs)
+
+        def test_numpy_equivalence(self):
+            lhs_array = np.arange(10)
+            rhs_array = np.arange(1, 11)
 
-    @abc.abstractmethod
-    def assertLessEqual(self, *args, **kwargs): ...
-
-    @abc.abstractmethod
-    def assertTrue(self, *args, **kwargs): ...
-
-    def test_equality(self):
-        DEFINITELY = 2
-        for x in self.generate_symbols():
-            self.assertEqual(DEFINITELY, x.maybe_equals(x))
-            self.assertTrue(x.equals(x))
-
-        for x, y in zip(self.generate_symbols(), self.generate_symbols()):
-            self.assertTrue(DEFINITELY, x.maybe_equals(y))
-            self.assertTrue(x.equals(y))
-
-    def test_inequality(self):
-        MAYBE = 1
-        for x, y in itertools.combinations(self.generate_symbols(), 2):
-            self.assertLessEqual(x.maybe_equals(y), MAYBE)
-            self.assertFalse(x.equals(y))
-
-    def test_iter_symbols(self):
-        for x in self.generate_symbols():
-            model = x.model
-            index = x.topological_index()
+            model = Model()
+            lhs_symbol = model.constant(lhs_array)
+            rhs_symbol = model.constant(rhs_array)
 
-            # Get the symbol back
-            y, = itertools.islice(model.iter_symbols(), index, index+1)
+            op_array = self.op(lhs_array, rhs_array)
+            op_symbol = self.symbol_op(lhs_symbol, rhs_symbol)
```

Comment:
Should there be another line here that actually tests the output is correct?

Reply:
:laughing: yes.

Reply:
Added in 369dc04